benchmarks/filtsmooth.py

```diff
@@ -0,0 +1,160 @@
+
+    def time_filter(self, linearization):
+        self.kalman_filter.filter(dataset=self.observations, times=self.locations)
+
+    def peakmem_filter(self, linearization):
+        self.kalman_filter.filter(dataset=self.observations, times=self.locations)
+
+
+class Smoothing:
+    """Benchmark Kalman filter for different linearization techniques."""
+
+    param_names = ["linearization"]
+    params = [["ekf", "ukf"]]
+
+    def setup(self, linearization):
+        dynmod, measmod, initrv, info = load_pendulum()
+        _lin_method = {
+            "ekf": pnfs.DiscreteEKFComponent,
+            "ukf": pnfs.DiscreteUKFComponent,
+        }[linearization]
+
+        self.locations = np.arange(0.0, info["tmax"], step=info["dt"])
+        _, self.observations = pnss.generate_samples(
+            dynmod=dynmod, measmod=measmod, initrv=initrv, times=self.locations
+        )
+
+        linearized_dynmod = _lin_method(dynmod)
+        linearized_measmod = _lin_method(measmod)
+
+        self.kalman_filter = pnfs.Kalman(
+            dynamics_model=linearized_dynmod,
+            measurement_model=linearized_measmod,
+            initrv=initrv,
+        )
+        self.filtering_posterior = self.kalman_filter.filter(
+            dataset=self.observations, times=self.locations
+        )
+
+    def time_smooth(self, linearization):
+        self.kalman_filter.smooth(filter_posterior=self.filtering_posterior)
+
+    def peakmem_smooth(self, linearization):
+        self.kalman_filter.smooth(filter_posterior=self.filtering_posterior)
+
+
+class Sampling:
+    """Benchmark Kalman filter for different linearization techniques."""
+
+    param_names = ["linearization", "num_samples"]
+    params = [["ekf", "ukf"], [1, 10]]
+
+    def setup(self, linearization, num_samples):
+        dynmod, measmod, initrv, info = load_pendulum()
+        _lin_method = {
+            "ekf": pnfs.DiscreteEKFComponent,
+            "ukf": pnfs.DiscreteUKFComponent,
+        }[linearization]
+
+        self.locations = np.arange(0.0, info["tmax"], step=info["dt"])
+        _, self.observations = pnss.generate_samples(
+            dynmod=dynmod, measmod=measmod, initrv=initrv, times=self.locations
+        )
+
+        linearized_dynmod = _lin_method(dynmod)
+        linearized_measmod = _lin_method(measmod)
+
+        self.kalman_filter = pnfs.Kalman(
+            dynamics_model=linearized_dynmod,
+            measurement_model=linearized_measmod,
+            initrv=initrv,
+        )
+        filtering_posterior = self.kalman_filter.filter(
+            dataset=self.observations, times=self.locations
+        )
+        self.smoothing_posterior = self.kalman_filter.smooth(
+            filter_posterior=filtering_posterior
+        )
+
+    def time_sample(self, linearization, num_samples):
+        self.smoothing_posterior.sample(size=num_samples)
+
+    def peakmem_sample(self, linearization, num_samples):
+        self.smoothing_posterior.sample(size=num_samples)
```

Comment:
Do you think it would make the sampling benchmark more meaningful if the sampling grid was more dense than the filtering grid?